File: parkaran-helper/api/topics.py
"""Topic counts, complete browsing, random starts and review evidence."""
import json
import random
import sqlite3
from contextlib import closing
from functools import lru_cache
from pathlib import Path
from flask import jsonify, request
import config
from database import corpus
from api.graph_api import graph_bp, _get_graph, _get_tag_vocab, _get_concepts, _get_sggs_sources


@lru_cache(maxsize=1)
def statistics():
    graph, concepts = _get_graph(), _get_concepts()
    with closing(sqlite3.connect((Path(config.DATA_DIR) / 'chroma_db/chroma.sqlite3').as_uri()+'?mode=ro', uri=True)) as db:
        denominator = db.execute("select count(distinct string_value) from embedding_metadata where key='shabad_id'").fetchone()[0]
    policy = _get_tag_vocab().get('generation_policy', {})
    denominator = policy.get('indexed_shabads', denominator)
    result = {}
    for tag, value in _get_tag_vocab().get('theme_tags', {}).items():
        assignments = [a for group in concepts.values() for a in group if a['concept'] == tag]
        lexical = sum(bool(a['lexical']) for a in assignments)
        inferred = len(assignments) - lexical
        result[tag] = {**value, 'tag': tag, 'count': len(graph['tag_index'].get(tag, [])),
                       'lexical': lexical, 'inferred': inferred, 'policy': policy,
                       'indexed_shabads': denominator,
                       'ak_count': sum(bool(_get_sggs_sources().get(sid, {}).get('amrit_keertan')) for sid in graph['tag_index'].get(tag, []))}
    return result


def eligible(tag, source):
    ids = _get_graph()['tag_index'].get(tag, [])
    if source == 'ak-only':
        ids = [sid for sid in ids if _get_sggs_sources().get(sid, {}).get('amrit_keertan')]
    return ids


def detail(sid):
    meta = _get_graph()['metadata'][sid]
    info = corpus.shabad(sid) or {}
    source = _get_sggs_sources().get(sid, {})
    return {'id': sid, 'gurmukhi': meta.get('gurmukhi', ''), 'title': meta.get('title', ''),
            'raag': info.get('raag', meta.get('raag', '')), 'writer': info.get('writer', ''),
            'ang': info.get('ang', meta.get('ang', 0)), 'is_amrit_keertan': bool(source.get('amrit_keertan')),
            'ak_chapters': source.get('ak_chapters', [])}


@graph_bp.route('/topics')
def topics():
    return jsonify(list(statistics().values()))


@graph_bp.route('/topics/<tag>')
def topic(tag):
    if tag not in statistics(): return jsonify({'error': 'Unknown topic'}), 404
    source = request.args.get('source', 'all')
    if source not in ('all', 'ak-only'): return jsonify({'error': 'Unknown source'}), 400
    ids = eligible(tag, source)
    offset = max(0, request.args.get('offset', 0, type=int))
    limit = max(1, min(50, request.args.get('limit', 20, type=int)))
    return jsonify({'topic': statistics()[tag], 'total': len(ids), 'offset': offset,
                    'next_offset': offset + limit if offset + limit < len(ids) else None,
                    'shabads': [detail(sid) for sid in ids[offset:offset+limit]]})


@graph_bp.route('/topics/<tag>/random')
def topic_random(tag):
    if tag not in statistics(): return jsonify({'error': 'Unknown topic'}), 404
    source = request.args.get('source', 'all')
    if source not in ('all', 'ak-only'): return jsonify({'error': 'Unknown source'}), 400
    ids = eligible(tag, source)
    excluded = set(request.args.get('exclude', '')[:2000].split(','))
    pool = [sid for sid in ids if sid not in excluded] or ids
    if not pool: return jsonify({'error': 'No shabads in this topic and source selection', 'total': 0}), 404
    return jsonify({'shabad': detail(random.SystemRandom().choice(pool)), 'total': len(ids), 'tag': tag})


@graph_bp.route('/topics/<tag>/review')
def topic_review(tag):
    if tag not in statistics(): return jsonify({'error': 'Unknown topic'}), 404
    review = json.loads((Path(config.DATA_DIR) / 'concept_review.json').read_text())
    record = review['concepts'][tag]
    reveal = request.args.get('reveal') == '1'
    samples = []
    for row in record['samples']:
        sample = {**detail(row['id']), 'line_index': row['line_index'], 'line_gurmukhi': row['line_gurmukhi']}
        if reveal: sample.update(category=row['category'], strength=row['strength'])
        samples.append(sample)
    manifest = json.loads((Path(config.DATA_DIR) / 'release-manifest.json').read_text())
    return jsonify({'topic': tag, 'dataset_id': manifest['dataset_id'], 'baseline_sha256': review['baseline_sha256'],
                    'note': 'Judge each example in its full shabad before revealing its previous assignment status. These selected examples do not estimate accuracy.',
                    'samples': samples, 'revealed': reveal})

File: parkaran-helper/static/js/topics.js
/* One topic panel for browsing, source-scoped random starts and expert review. */
const Topic = {all: null, tag: null, source: 'all', offset: 0, generation: 0, recent: {}};

async function openTagBrowser(tag = null) {
    hideTooltip();
    const modal = document.getElementById('topicModal');
    modal.classList.remove('hidden');
    try {
        if (!Topic.all) Topic.all = await API.get('/api/topics');
        renderTopicIndex();
        if (tag) await selectTag(tag);
    } catch (error) {
        document.getElementById('topicIndex').textContent = 'Topics unavailable. Close and try again.';
    }
}
function closeTagBrowser() {
    Topic.generation++;
    document.getElementById('topicModal').classList.add('hidden');
}
function openTagShabadsModal(tag) { return openTagBrowser(tag); }
function closeTagShabadsModal() { closeTagBrowser(); }

function renderTopicIndex() {
    const query = document.getElementById('topicSearch').value.trim().toLocaleLowerCase();
    const topics = (Topic.all || []).filter(t => [t.tag,t.gurbani_term,t.description].join(' ').toLocaleLowerCase().includes(query));
    const index = document.getElementById('topicIndex');
    index.innerHTML = topics.map(t => `<button type="button" class="topic-chip" data-tag="${escAttr(t.tag)}" aria-pressed="${Topic.tag === t.tag}">${escapeHtml(t.tag)} <small>${t.count.toLocaleString()}</small></button>`).join('') || '<p>No matching topics.</p>';
    index.querySelectorAll('[data-tag]').forEach(button => button.onclick = () => selectTag(button.dataset.tag));
}
async function selectTag(tag, offset = 0) {
    Topic.tag = tag; Topic.offset = offset;
    const generation = ++Topic.generation;
    const source = document.getElementById('topicSource').value;
    Topic.source = source;
    renderTopicIndex();
    const detail = document.getElementById('topicDetail');
    detail.innerHTML = '<p role="status">Loading topic…</p>';
    try {
        const data = await API.get(`/api/topics/${encodeURIComponent(tag)}?offset=${offset}&source=${source}`);
        if (generation !== Topic.generation) return;
        const t = data.topic;
        detail.innerHTML = `<header class="topic-heading"><div><span class="eyebrow">Explore a topic</span><h2>${escapeHtml(tag)}</h2></div><button type="button" id="topicRandom" ${data.total ? '' : 'disabled'}>Random starting shabad</button></header>
            <p>${escapeHtml(t.description)}</p><p class="topic-counts"><strong>${t.count.toLocaleString()}</strong> shabads · ${t.lexical.toLocaleString()} lexical · ${t.inferred.toLocaleString()} inferred · ${t.ak_count.toLocaleString()} in Amrit Keertan</p>
            ${t.confidence === 'description-only' ? '<p class="connection-notice">This topic currently uses description-only matching. All its assignments are inferred.</p>' : ''}
            <details class="topic-method"><summary>How this count was produced</summary><p>Every lexical match is retained. Inferred assignments use a separate similarity threshold for this concept, without a percentage ceiling or a multiple of lexical frequency. The current threshold is automatically calibrated and remains provisional. A lexical occurrence also needs interpretation in context.</p><p>Previous dataset: ${Number(t.previous_count || 0).toLocaleString()} assignments. Newly admitted inferences: ${Number(t.new_inferred || 0).toLocaleString()}. Graph display limits do not change topic membership.</p></details>
            <div class="topic-tools"><button type="button" id="topicReview">Review assignment examples</button><span>${data.total ? `${offset+1}–${Math.min(offset+20,data.total)} of ${data.total.toLocaleString()}` : 'No shabads in this source selection'}</span></div>
            <div id="topicReviewSamples"></div><div class="topic-results">${data.shabads.map(s => `<article class="topic-row"><button type="button" class="topic-open" data-open="${s.id}"><span lang="pa-Guru">${escapeHtml(s.gurmukhi || s.title)}</span><small>${escapeHtml([s.raag,s.writer,'Ang '+s.ang].filter(Boolean).join(' · '))}</small></button>${sourceBadge(s)}<button type="button" data-read="${s.id}">Read</button></article>`).join('')}</div>
            <nav class="topic-pagination"><button type="button" id="topicPrevious" ${offset ? '' : 'disabled'}>Previous</button><button type="button" id="topicNext" ${data.next_offset === null ? 'disabled' : ''}>Next</button></nav>`;
        document.getElementById('topicRandom').onclick = () => randomTopic(tag, source);
        document.getElementById('topicPrevious').onclick = () => selectTag(tag, Math.max(0,offset-20));
        document.getElementById('topicNext').onclick = () => selectTag(tag, data.next_offset);
        document.getElementById('topicReview').onclick = () => reviewTopic(tag, generation);
        detail.querySelectorAll('[data-open]').forEach(b => b.onclick = () => startFromTopic(tag,b.dataset.open));
        detail.querySelectorAll('[data-read]').forEach(b => b.onclick = () => loadPreview(b.dataset.read));
    } catch (error) {
        if (generation === Topic.generation) detail.innerHTML = '<p>Could not load this topic. Choose it again to retry.</p>';
    }
}
async function randomTopic(tag, source) {
    const key = tag+'|'+source;
    const recent = Topic.recent[key] || [];
    const button = document.getElementById('topicRandom');
    if (button) button.disabled = true;
    const generation = ++Topic.generation;
    try {
        const data = await API.get(`/api/topics/${encodeURIComponent(tag)}/random?source=${source}&exclude=${recent.join(',')}`);
        if (generation !== Topic.generation) return;
        Topic.recent[key] = [...recent.slice(-29),data.shabad.id];
        State.metadata[data.shabad.id] = {...State.metadata[data.shabad.id],...data.shabad};
        startFromTopic(tag,data.shabad.id,source);
    } catch (error) { showToast('No starting shabad available in this topic and source selection.'); }
    finally { if (button) button.disabled = false; }
}
function startFromTopic(tag, sid, source = Topic.source) {
    closeTagBrowser();
    switchToTab('explore');
    delete State.selectedTuk[sid];
    State.startingTopic = {tag,source};
    document.getElementById('topicOrigin').classList.remove('hidden');
    document.getElementById('topicOriginLabel').textContent = `Started from ${tag}`;
    document.getElementById('topicAnother').onclick = () => randomTopic(tag,source);
    expandShabad(sid);
}
function surpriseMe() {
    const ids = Object.keys(State.metadata);
    const other = ids.filter(id => id !== State.centerNode);
    if (!other.length) return;
    State.startingTopic = null;
    document.getElementById('topicOrigin').classList.add('hidden');
    const sid = other[Math.floor(Math.random()*other.length)];
    delete State.selectedTuk[sid];
    expandShabad(sid);
}
async function reviewTopic(tag, generation, reveal = false) {
    const box = document.getElementById('topicReviewSamples');
    box.innerHTML = '<p>Loading evidence…</p>';
    try {
        const data = await API.get(`/api/topics/${encodeURIComponent(tag)}/review?reveal=${reveal ? 1 : 0}`);
        if (generation !== Topic.generation) return;
        const key = `shabadverse_review:${data.dataset_id}:${tag}`;
        let judgments = {};
        if (reveal) localStorage.setItem(key+':revealed','1');
        const hasRevealed = localStorage.getItem(key+':revealed') === '1';
        try { judgments = JSON.parse(localStorage.getItem(key) || '{}'); } catch (_) {}
        const labels = {lexical:'Lexical example',previously_retained:'Previously retained inference',newly_admitted:'Newly admitted inference'};
        box.innerHTML = `<p>${escapeHtml(data.note)}</p><p>Judgments stay in this browser. Export them to preserve a review record.</p>${data.samples.map((s,i)=>`<article><p>Example ${i+1}${reveal ? ' · '+escapeHtml(labels[s.category] || s.category) : ''}</p><p lang="pa-Guru">${escapeHtml(s.line_gurmukhi)}</p><small>Shabad ${s.id} · verse ${s.line_index+1}</small> <button type="button" data-read="${s.id}">Read full shabad</button><label>Your judgment <select data-judge="${s.id}"><option value="">Unreviewed</option><option value="supported">Supported in context</option><option value="unsupported">Not supported in context</option><option value="uncertain">Uncertain</option></select></label></article>`).join('')}<button type="button" id="revealTopicReview" ${reveal ? 'disabled' : ''}>Reveal previous status</button> <button type="button" id="exportTopicReview">Export judgments</button>`;
        box.querySelectorAll('[data-read]').forEach(b=>b.onclick=()=>loadPreview(b.dataset.read));
        box.querySelectorAll('[data-judge]').forEach(select=>{
            select.value=judgments[select.dataset.judge]?.judgment || '';
            select.onchange=()=>{
                judgments[select.dataset.judge]={judgment:select.value,after_reveal:hasRevealed,recorded_at:new Date().toISOString()};
                try { localStorage.setItem(key,JSON.stringify(judgments)); } catch (_) { showToast('Could not save this judgment. Export before closing.'); }
            };
        });
        document.getElementById('revealTopicReview').onclick=()=>reviewTopic(tag,generation,true);
        document.getElementById('exportTopicReview').onclick=()=>{
            const record={dataset_id:data.dataset_id,baseline_sha256:data.baseline_sha256,topic:tag,judgments,samples:data.samples};
            const url=URL.createObjectURL(new Blob([JSON.stringify(record,null,2)],{type:'application/json'}));
            const a=document.createElement('a');a.href=url;a.download=`shabadverse-review-${tag}.json`;a.click();setTimeout(()=>URL.revokeObjectURL(url),1000);
        };
    } catch(error) { if(generation === Topic.generation) box.textContent='Evidence unavailable. Try again.'; }
}
